lib/HTFeed/StorageZipAudit.pm

```diff
@@ -0,0 +1,199 @@
+    my $config = get_config('storage_classes');
+    my $storage_config = $config->{$storage_name};
+    die "unable to get config for $storage_name" unless $storage_config;
+
+    my @components = split '::', $storage_config->{class};
+    my $module_class = $class . '::'. $components[-1];
+    my $module_path = $module_class . '.pm';
+    $module_path =~ s/::/\//g;
+    require $module_path;
+    my $self = $module_class->new();
+    $self->{storage_name} = $storage_name;
+    return $self;
+  }
+  else {
+    return bless({}, $class);
+  }
+}
+
+# Choose and if necessary request a random object for this storage class.
+sub random_object {
+  my $self = shift;
+
+  my ($namespace, $objid, $path, $version);
+  my $sql = 'SELECT namespace,id,path,version FROM feed_backups' .
+            ' WHERE storage_name=?' .
+            ' AND deleted IS NULL' .
+            ' AND saved_md5sum IS NOT NULL' .
+            ' AND restore_request IS NULL' .
+            ' ORDER BY RAND() LIMIT 1';
+  eval {
+    my $ref = get_dbh->selectall_arrayref($sql, undef, $self->{storage_name});
+    ($namespace, $objid, $path, $version) = @{$ref->[0]} if scalar @$ref;
+  };
+  if ($@) {
+    die "Database query failed: $@";
+  }
+  return unless $namespace && $objid && $path && $version;
+  return $self->storage_object($namespace, $objid, $path, $version);
+}
+
+
+# Get the object(s) selected for audit by choose_random_object() or some other criteria.
+# In the case of AWS Glacier, it is one or more objects selected in a previous run
+# that have been restored in the intervening time.
+# These are the objects on the filesystem ready for checksumming.
+sub all_objects {
+  my $self = shift;
+
+  return [$self->random_object()];
+}
+
+sub storage_object {
+  my $self      = shift;
+  my $namespace = shift;
+  my $objid     = shift;
+  my $path      = shift;
+  my $version   = shift;
+  
+  unless ($namespace && $objid && $path && $version) {
+    croak "invalid storage_object() args: namespace, objid, path, version required";
+  }
+  my $obj = {namespace => $namespace, objid => $objid, version => $version,
+             path => $path, storage_name => $self->{storage_name}};
+  $obj->{volume} = HTFeed::Volume->new(packagetype => 'pkgtype', namespace => $namespace, objid => $objid);
+  my $config = get_config('storage_classes');
+  $obj->{storage_config} = $config->{$self->{storage_name}};
+  $obj->{storage} = $obj->{storage_config}->{class}->new(volume => $obj->{volume},
+                                                         config => $obj->{storage_config},
+                                                         name   => $self->{storage_name});
+  $obj->{storage}->{zip_suffix} = '.gpg';
+  $obj->{storage}->{timestamp} = $obj->{version};
+  $obj->{zip_file} = $obj->{storage}->zip_filename();
+  $obj->{mets_file} = $obj->{storage}->mets_filename();
+  $obj->{zip_path} = $self->storage_object_path($obj) . '/' . $obj->{zip_file};
```

Comment:
Hmmm.. Perhaps we should consider using `$obj->{storage}->zip_obj_path` / `$obj->{storage}->mets_obj_path` and asserting that those are the same as what's in the database table?

Reply:
I may be misunderstanding your comment, but this path juggling is specifically for the Glacier case where we make a temp location for the downloaded files. I mean, we could make that assertion for any class other than ObjectStore. Maybe I should call it ephemeral_object_path.

Reply:
Right -- I think I didn't quite understand that at first glance here,  and was thinking primarily about the Data Den case. It's probably worth thinking about how to isolate the special behavior to ObjectStore, but I think that's pretty low priority for right now.

Reply:
I can get rid of the redundant `zip_file` and `mets_file` attrs regardless.

Reply:
The way I rationalize it, the storage class need not know where to "download" things. ObjectStore is just weird.